KAFKA-10282: Remove Log metrics immediately when deleting log

bob-barrett opened the pull request:
Currently, we remove the Log metrics when asynchronous deletion of the log is triggered. However, we attempt to register the metrics immediately upon log creation. If a Log object is re-created for a partition that is pending deletion (because a topic was quickly re-created or because a partition was moved off and back onto a broker), the registration of the new metrics can happen before the asyncrhonous deletion. In this case, the metrics are removed after the second registration, leading to missing Log metrics.

To fix this, this patch changes the log deletion behavior to remove the metrics when the log is first marked for deletion, rather than when the files are deleted. This removes the window in which metrics registration can occur before metrics removal. This is justifiable because the log should be logically deleted when a delete request or partition movement finishes, rather than when the files are actually removed. Tested with unit tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

bob-barrett committed:
KAFKA-10282: Remove Log metrics immediately when deleting log

Currently, we remove the Log metrics when asynchronous deletion of the log is triggered. However, we attempt to register the metrics immediately upon log creation. If a Log object is re-created for a partition that is pending deletion (because a topic was quickly re-created or because a partition was moved off and back onto a broker), the registration of the new metrics can happen before the asyncrhonous deletion. In this case, the metrics are removed after the second registration, leading to missing Log metrics.

To fix this, this patch changes the log deletion behavior to remove the metrics when the log is first marked for deletion, rather than when the files are deleted. This removes the window in which metrics registration can occur before metrics removal. This is justifiable because the log should be logically deleted when a delete request or partition movement finishes, rather than when the files are actually removed. Tested with unit tests.

dajac reviewed (approved):
LGTM, thanks for the fix.

dhruvilshah3 commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -978,6 +978,7 @@ class LogManager(logDirs: Seq[File],
         checkpointRecoveryOffsetsAndCleanSnapshotsInDir(logDir, logsToCheckpoint, ArrayBuffer.empty)
         checkpointLogStartOffsetsInDir(logDir, logsToCheckpoint)
       }
+      removedLog.removeLogMetrics()
```
We use the same pattern of removing the log from `futureLogs` or `currentLogs` and then removing the corresponding log metrics in `handleLogDirFailure`. Do you think it'd be useful to turn this into a helper method and reuse the logic in both places?

bob-barrett replied:
That's a good catch, thanks. I'll add that.

dhruvilshah3 commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -563,4 +565,39 @@ class LogManagerTest {
+      assertTrue(metric.getMBeanName.contains(metricTag))
+    }
+
+    time.sleep(logConfig.fileDeleteDelayMs + 1)
```
Why is this sleep needed?

bob-barrett replied:
It causes enough "time" to elapse that the actual file deletion is triggered. Having it here tests that we still have metrics after the deletion occurs. I'll add a comment to that effect.

soondenana replied:
Is it deterministic? I realize its mocktime, but threads can be scheduled in any order. Or using `MockTime` means all schedulers will get triggered before next line in test will get to run?

bob-barrett replied:
Yes, `MockTime` contains a `MockScheduler`, which explicitly checks for any scheduled tasks that need to be run whenever we sleep. And we pass that `MockScheduler` into the `Log` and `LogManager` constructors in the tests. See https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/utils/MockScheduler.scala#L59.

soondenana commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -563,4 +565,39 @@ class LogManagerTest {
+  def testMetricsExistWhenLogIsRecreatedBeforeDeletion(): Unit = {
+    val topicName = "metric-test"
+    def logMetrics = KafkaYammerMetrics.defaultRegistry.allMetrics.keySet.asScala.
+      filter(_.getType == "Log").filter(_.getScope.contains(topicName))
```
nit: maybe combine consecutive filters.

soondenana commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -563,4 +565,39 @@ class LogManagerTest {
+
+    logManager.getOrCreateLog(tp, () => logConfig)
+
+    assertEquals(LogMetricNames.allMetricNames.size, logMetrics.size)
+    logMetrics.foreach { metric =>
+      assertTrue(metric.getMBeanName.contains(metricTag))
+    }
```
duplicate code, its being scala maybe define an internal helper method. There is another instance of this below.

soondenana reviewed (commented):
Thanks for fixing it. Don't have much to add to what @dhruvilshah3 said. Left couple of comments in test.

bob-barrett committed:
PR feedback, handle future logs

bob-barrett commented:
While addressing the feedback, I realized I had missed the case when replacing a current log with a future log as part of altering a partition's log dir. We also deferred metrics removal until the file deletion in that case. That has been fixed now.

soondenana commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1151,6 +1144,12 @@ class LogManager(logDirs: Seq[File],
+  private def removeLogAndMetrics(logs: Pool[TopicPartition, Log], tp: TopicPartition): Log = {
+    val removedLog = logs.remove(tp)
+    if (removedLog != null) removedLog.removeLogMetrics()
+    removedLog
```
nit: Lets return Option(removedLog) to ease null checking by clients.

Seems like the same object gets returned by `asyncDelete` but is only used in one place in test code, so many want to change the return value of that too. The less "null" we have the better.

soondenana commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -199,27 +199,22 @@ class LogManager(logDirs: Seq[File],
-        if (removedLog != null) {
-          removedLog.closeHandlers()
-          removedLog.removeLogMetrics()
+      def removeOfflineLogs(logs: Pool[TopicPartition, Log]): Iterable[TopicPartition] = {
```
Thanks for deduping this code.

soondenana reviewed (approved):
LGTM

Left one nit.

bob-barrett committed:
Make removeLogAndMetrics return an Option[Log]

soondenana commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1151,6 +1149,16 @@ class LogManager(logDirs: Seq[File],
+    val removedLog = logs.remove(tp)
+    if (removedLog != null) {
+      removedLog.removeLogMetrics()
+      Some(removedLog)
+    } else {
+      None
+    }
```
nit: I think just returning `Option(removedLog)` should take care of Some/None part.

soondenana commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -199,27 +199,22 @@ class LogManager(logDirs: Seq[File],
-      }}
+        offlineTopicPartitions.foreach { topicPartition => {
+          val removedLog = removeLogAndMetrics(logs, topicPartition)
+          removedLog.foreach {
+            log => log.closeHandlers()
+          }
```
nit: as its one line, can be abbreviated as `_.closeHandlers()`

soondenana reviewed (approved):
Thanks for fixing it.

Couple more nits, feel free to ignore if its too much work (or you don't see value)

ijuma commented:
ok to test

dhruvilshah3 commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -932,6 +927,7 @@ class LogManager(logDirs: Seq[File],
         val logsToCheckpoint = logsInDir(logDir)
         checkpointRecoveryOffsetsAndCleanSnapshotsInDir(logDir, logsToCheckpoint, ArrayBuffer.empty)
         checkpointLogStartOffsetsInDir(logDir, logsToCheckpoint)
+        sourceLog.removeLogMetrics()
```
This seems a bit odd but looks like an existing issue. We are removing metrics for the `current` log but we leave the metrics for the `future` log as is, which means it will retain the `is-future` tag. Maybe worth filing a JIRA if you think it is a problem too.

bob-barrett replied:
Good catch. Opened https://issues.apache.org/jira/browse/KAFKA-10320 to track it.

dhruvilshah3 reviewed (approved):
Thanks for the PR, LGTM. I left a comment about a potential existing issue.

gwenshap commented:
retest this please

gwenshap commented:
Test failures:
org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.testOneWayReplicationWithAutorOffsetSync1
in both JDK 11 and JDK 14. My understanding is that this isn't related to this PR.

bob-barrett commented:
Test failure looks unrelated and passes locally, probably a flake

gwenshap reviewed (commented):
post-facto LGTM 👍
